components/server/src/oauth-server/oauth-controller.ts

```diff
@@ -0,0 +1,148 @@
+/**
+ * Copyright (c) 2020 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+import { AuthCodeRepositoryDB } from '@gitpod/gitpod-db/lib/typeorm/auth-code-repository-db';
+import { UserDB } from '@gitpod/gitpod-db/lib/user-db';
+import { User } from "@gitpod/gitpod-protocol";
+import { log } from '@gitpod/gitpod-protocol/lib/util/logging';
+import { OAuthException, OAuthRequest, OAuthResponse } from "@jmondi/oauth2-server";
+import * as express from 'express';
+import { inject, injectable } from "inversify";
+import { Env } from "../env";
+import { createAuthorizationServer } from './oauth-authorization-server';
+import { localAppClientID } from "./db";
+
+@injectable()
+export class OAuthController {
+    @inject(Env) protected readonly env: Env;
+    @inject(UserDB) protected readonly userDb: UserDB;
+    @inject(AuthCodeRepositoryDB) protected readonly authCodeRepositoryDb: AuthCodeRepositoryDB;
+
+    get oauthRouter(): express.Router {
+        const router = express.Router();
+        if (this.env.enableOAuthServer) {
+            const authorizationServer = createAuthorizationServer(this.authCodeRepositoryDb, this.userDb, this.userDb);
+            router.get("/oauth/authorize", async (req: express.Request, res: express.Response) => {
+                log.info(`AUTHORIZE: ${JSON.stringify(req.query)}`);
+
+                if (!req.isAuthenticated() || !User.is(req.user)) {
+                    const redirectTarget = encodeURIComponent(`${this.env.hostUrl}api${req.originalUrl}`);
+                    const redirectTo = `${this.env.hostUrl}login?returnTo=${redirectTarget}`;
+                    log.info(`AUTH Redirecting to login: ${redirectTo}`);
+                    res.redirect(redirectTo)
+                    return
+                }
+
+                const user = req.user as User;
+                if (user.blocked) {
+                    res.sendStatus(403);
+                    return;
+                }
+
+                // Have they authorized the local-app?
+                const wasApproved = req.query['approved'] || '';
```

Comment:
Not so much a shortcut, but yes it is part of the web flow, in particular the rejection side so it returns immediately.

It is probably better to keep the explicit rejection in the db as opposed to deleting the entry. That way we can differentiate between not having asked for approval and an explicit rejection. The downside is that if they've changed their mind after initially rejecting it we have no easy way for them to do that in the UI... yet. It can be done via an explicitly formed URL but is not user-friendly.